src/app/community-list-page/community-list-service.ts

```diff
@@ -0,0 +1,288 @@
+import {Injectable} from '@angular/core';
+import {combineLatest as observableCombineLatest} from 'rxjs/internal/observable/combineLatest';
+import {Observable, of as observableOf} from 'rxjs';
+import {CommunityDataService} from '../core/data/community-data.service';
+import {PaginationComponentOptions} from '../shared/pagination/pagination-component-options.model';
+import {SortDirection, SortOptions} from '../core/cache/models/sort-options.model';
+import {catchError, filter, map, switchMap, take} from 'rxjs/operators';
+import {Community} from '../core/shared/community.model';
+import {Collection} from '../core/shared/collection.model';
+import {hasValue, isNotEmpty} from '../shared/empty.util';
+import {RemoteData} from '../core/data/remote-data';
+import {PaginatedList} from '../core/data/paginated-list';
+import {getCommunityPageRoute} from '../+community-page/community-page-routing.module';
+import {getCollectionPageRoute} from '../+collection-page/collection-page-routing.module';
+import {CollectionDataService} from '../core/data/collection-data.service';
+
+export interface FlatNode {
+    isExpandable$: Observable<boolean>;
+    name: string;
+    id: string;
+    level: number;
+    isExpanded?: boolean;
+    parent?: FlatNode;
+    payload: Community | Collection | ShowMoreFlatNode;
+    isShowMoreNode: boolean;
+    route?: string;
+    currentCommunityPage?: number;
+    currentCollectionPage?: number;
+}
+
+export class ShowMoreFlatNode {
+}
+
+export const combineAndFlatten = (obsList: Array<Observable<FlatNode[]>>): Observable<FlatNode[]> =>
+    observableCombineLatest(...obsList).pipe(
+        map((matrix: FlatNode[][]) =>
+            matrix.reduce((combinedList, currentList: FlatNode[]) => [...combinedList, ...currentList]))
+    );
+
+export const toFlatNode = (
+    c: Community | Collection,
+    isExpandable: Observable<boolean>,
+    level: number,
+    isExpanded: boolean,
+    parent?: FlatNode
+): FlatNode => ({
+    isExpandable$: isExpandable,
+    name: c.name,
+    id: c.id,
+    level: level,
+    isExpanded,
+    parent,
+    payload: c,
+    isShowMoreNode: false,
+    route: c instanceof Community ? getCommunityPageRoute(c.id) : getCollectionPageRoute(c.id),
+});
+
+export const showMoreFlatNode = (
+    id: string,
+    level: number,
+    parent: FlatNode
+): FlatNode => ({
+    isExpandable$: observableOf(false),
+    name: 'Show More Flatnode',
+    id: id,
+    level: level,
+    isExpanded: false,
+    parent: parent,
+    payload: new ShowMoreFlatNode(),
+    isShowMoreNode: true,
+});
+
+// tslint:disable-next-line:max-classes-per-file
+@Injectable()
+export class CommunityListService {
```

Comment:
Again, TypeDocs please.  I'd also appreciate comments to describe the various exported `const` above.